compiler/damlc/tests/daml-test-files/InterfaceGuarded.daml

```diff
@@ -0,0 +1,91 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- @SINCE-LF-FEATURE DAML_INTERFACE
+
+module InterfaceGuarded where
+
+import DA.Exception (GeneralError (..), throwPure)
+
+interface Token where
+  getOwner : Party
+  getAmount : Int
+  setAmount : Int -> Token
+
+  ensure (getAmount this >= 0)
+
+  choice GetRich : ContractId Token
+    with
+      byHowMuch : Int
+    controller getOwner this
+    do
+        assert (byHowMuch > 0)
+        create $ setAmount this (getAmount this + byHowMuch)
+
+template Asset
+  with
+    issuer : Party
+    owner : Party
+    amount : Int
+  where
+    signatory owner
+
+    implements Token where
+      let getOwner = owner
+      let getAmount = amount
+      let setAmount = \x -> toInterface @Token (this with amount = x)
+
+exception GuardException
+    with
+        m : Text
+    where
+        message m
+
+main = scenario do
+  p <- getParty "Alice"
+  let
+    assetTpl = Asset with
+      issuer = p
+      owner = p
+      amount = 100
+    getRich = GetRich with byHowMuch = 1000000
+
+  p `submit` do
+    -- OK if guard returns True
+    asset <- create assetTpl
+    exerciseGuarded (const True) asset getRich
+
+  p `submitMustFail` do
+    -- Fail if guard returns False
+    asset <- create assetTpl
+    exerciseGuarded (const False) asset getRich
+
+  p `submitMustFail` do
+    -- Fail if guard throws exception
+    asset <- create assetTpl
+    exerciseGuarded (\_ -> error "foo") asset getRich
+
+  p `submitMustFail` do
+    -- Fail if guard throws custom exception
+    asset <- create assetTpl
+    exerciseGuarded (\_ -> throwPure (GuardException "bar")) asset getRich
+
+  p `submitMustFail` do
```

Comment:
@cocreature I thought I would need some extra work for this and the following `submitMustFail` to indeed fail, but they seem to do the right thing - neither `error "foo"` nor `GuardException "bar"` are caught. 

I tried running these with `submit` instead of `submitMustFail` to see the errors, and they both are

```
      Message:
        Scenario execution failed on commit at InterfaceGuarded:73:3:
        CRASH: ChoiceGuardFailed unhandled in scenario service
```

so perhaps we want to do something about that?

Reply:
ah yes I saw that one, I meant to say that I'm not sure if we get the right behavior "by accident" and fixing that todo might break this test. That is, I couldn't convince myself that errors thrown in the predicate will always be uncatchable

Reply:
Yeah, `ChoiceGuardFailed` is currently not handled properly in the scenario service. The same is true for `ContractDoesNotImplementInterface`. I would treat these separately.

Reply:
The `ChoiceGuardFailed` is the right behavior :) An exception is uncatcheable because it's turned into `ChoiceGuardFailed`.

Reply:
awesome, thanks for clearing that up for me!